### ISSUE TITLE
Consider changes in MatrixSSL config in Makefile.PL

### ISSUE THREAD

**Issue by powerman** [powerman/perl-Crypt-MatrixSSL3#5]
Looks like many features from commercial version now available in open source version.

- enable client certificates
- enable most/all ciphers
  - from one point it makes a lot of sense to disable weak ciphers by default
  - from another this is general module and user should be able to use any ciphers if needed
  - maybe solution is to enable all ciphers but also provide new constant with list of recommended ciphers which can be easily used to setup new connections?
- either DISABLE_SSLV3 to make it more secure, or, better, let user choose allowed protocols (will need to add ability to provide `options` param which is now absent from perl API)

**Comment by vonydev** [comment]
>> enable client certificates - just to be sure: you are referring to client authentication? If yes I added
the 'USE_CLIENT_AUTH' define in Makefile.PL. Should this be enabled by default?
>> enable most/all ciphers  - I don't see a reason for this. The current cipher selection includes the
most common ciphers. If somebody needs older/not_so_used ciphers they can edit the Makefile.PL
and enabled what's needed. Alternatively we can add some prompts/defines to enable them when
the makefile is generated
>> either DISABLE_SSLV3 - SSL3 is disabled by default. TLS 1.0/1.1/1.2 enabled by default

**Comment by powerman** [comment]
> just to be sure: you are referring to client authentication?

Yeah, it's about USE_CLIENT_AUTH.

> If somebody needs older/not_so_used ciphers they can edit the Makefile.PL and enabled what's needed.
> Alternatively we can add some prompts/defines to enable them when
the makefile is generated.

Why we should ask users to do this extra work instead of just `cpanm Crypt::MatrixSSL3` and have everything they may need? What is downside of compiling support for all ciphers by default and then provide one constant with predefined "recommended & secure" set of ciphers?

> SSL3 is disabled by default.

What do you mean? Makefile.PL doesn't set 'DISABLE_SSLV3', so I suppose it's enabled. Or you mean when using default (zeroed) `options` param then SSLv3 won't be allowed? If yes then it's ok for me to keep it this way - actually it's consistent with previous item (compiling support for all ciphers, even not secure ones, just like SSLv3).

**Comment by vonydev** [comment]
> Why we should ask users to do this extra work instead of just cpanm Crypt::MatrixSSL3 and have everything they may need? What is downside of compiling support for all ciphers by default and then provide one constant with predefined "recommended & secure" set of ciphers?

All compiled ciphers are enabled by default - this is how matrixSSL is built. For a secure 'out-of-the-box' usage we will need to disable the weak ones at start (after matrixSslOpen() is called) and provide a method for user to enable them on request.

> Makefile.PL doesn't set 'DISABLE_SSLV3'

It's disabled by default in the official matrixSSL distribution. That setting doesn't have any effect anymore. I'll remove it.

**Comment by powerman** [comment]
> For a secure 'out-of-the-box' usage we will need to disable the weak ones at start (after matrixSslOpen() is called) and provide a method for user to enable them on request.

Ah, I see, you're talking about server setup and `matrixSslSetCipherSuiteEnabledStatus()`. How about this:

- all ciphers are enabled at compile time
- add constant `DEFAULT_CIPHERS` set to ARRAYREF with recommended secure ciphers
- if user calls `Crypt::MatrixSSL3::Client->new()` with undef in `\@cipherSuites` param then `DEFAULT_CIPHERS` will be used instead
  - user can make it obvious by using `Crypt::MatrixSSL3::Client->new(…, DEFAULT_CIPHERS, …)`
- after `matrixSslOpen()` is called (no matter how - from `add_obj()` or manually by user as `Open()`) we will automatically call `matrixSslSetCipherSuiteEnabledStatus()` to disable all ciphers not included in DEFAULT_CIPHERS
  - this way server will automatically have same setup as now
  - also this will be better than current implementation in case user will uncomment ciphers in Makefile.PL - because now this will result in enabling these ciphers both for server and client, while user may need them only for client
- we probably will have to change `#define SSL_MAX_DISABLED_CIPHERS 8` to larger value (now we've 54 ciphers and about 34 disabled by default, so we need to set SSL_MAX_DISABLED_CIPHERS to 48 at least

The downsides are:
- about 34 extra calls to `matrixSslSetCipherSuiteEnabledStatus()` each time `matrixSslOpen()` is called
- few more bytes for `uint16 disabledCiphers[SSL_MAX_DISABLED_CIPHERS];` in MatrixSSL
- few more loops when MatrixSSL does `for (j = 0; j < SSL_MAX_DISABLED_CIPHERS; j++) {`
- a bit more XS code to handle `DEFAULT_CIPHERS`

I think these downsides are not important and user's ability to use any ciphers without manual editing Makefile.PL and recompilation of module is worth it.

**Comment by vonydev** [comment]
> we probably will have to change #define SSL_MAX_DISABLED_CIPHERS 8 to larger value (now we've 54 ciphers and about 34 disabled by default, so we need to set SSL_MAX_DISABLED_CIPHERS to 48 at least

SSL_MAX_DISABLED_CIPHERS is only relevant when enabling/disabling ciphers per session. When matrixSslSetCipherSuiteEnabledStatus() is called with a NULL session, this means enable/disable ciphers globally and this limit is not used. 

The 34 extra calls are irrelevant because this is done only once.

The other downsides are non existent if someone just modifies the Makefile.PL and comments weak ciphers so they don't get included by default. There is always this option.

**Comment by powerman** [comment]
Okay. I'll assign this to you. You're free to implement this, reject, or ask me to help with implementation if needed. There are no urgent changes critical for next release on CPAN in this issue, but it would be nice to have, of course.
